### DISCUSSION TITLE
Where is NRP[N] Logic in Codebase? Specifically the [-2,-1] handling (token[5] maybe?) [CC,ByteValue,MSB7bitValue,99,`-2`]:

### DISCUSSION THREAD

**Discussion by unityconstruct** [RomanKubiak/ctrlr#610] [Q&A] [answered]
### Where is the handing for this MidiMessageMultiList in the code? JUCE? Looking for Classname/MethodName
- I think the logic checks token[5] & if [-2] or [-1] DON'T do something that you would otherwise.
- I know what it does & that it works, but NEED to see the logic so I can document & explain it for various targets.

```
CC,ByteValue,MSB7bitValue,99,`-2`
CC,ByteValue,LSB7bitValue,98,`-2`
CC,ByteValue,LSB7bitValue,6,`-1`
```

Thanks in advance for any responses!... 
~uc

**Comment by Godlike-Productions** [comment]
I think the logic will be in the Ctrlr editor panel, not MIDI.  As far as LUA is concerned this is just sending a MIDI Message.  That will be bound to Juce using Luabind.

**Comment by Godlike-Productions** [comment]
NRPN's are defined after line 664 of https://github.com/damiensellier/CtrlrX/blob/def8d6dcd72f90d71a8d90450c706a9bb669a56c/Source/Resources/XML/CtrlrIDs.xml#L146

NRPN's aren't that complicated.  They are simply 4 MIDI CC's.

Wikipedia explains them quite well. https://en.wikipedia.org/wiki/NRPN

Basically send MIDI CC98 and then MIDICC99 to pick what parameter you are controlling then
Send Data (MIDI CC6), then MIDI CC38 for the value.  CC6 is MSB, 38 is LSB if 14 bit control range is available.

As to what parameter CC98 and 99 this is manufacturer and synth dependent, similar to the sysex spec.  You'll need to track down a manual or document from the manufacturer to decode these for each synth.  Note not every syth supports these.

**Comment by dobo365** [comment]
@unityconstruct, can you please try to be a bit less verbose (not only in this thread but in all)?
I don't think a forum is the place to write all our possible thoughts (all of use have many thoughts and are searching a lot).
Better be concise. Further readers of threads don't want to spend hours in looking for some simple answer to a precise question.
Thx for that.

@Godlike-Productions answered you on where the definitions are.

I'm completing that by saying that the last attribute -2 or -1 is used to indicate "controller" (-2) or "value" (-1)
So, for a MultiMessage, like

```
CC,ByteValue,MSB7bitValue,99,-2
CC,ByteValue,LSB7bitValue,98,-2
CC,ByteValue,MSB7bitValue,6,-1
CC,ByteValue,LSB7bitValue,38,-1
```

The 2 first lines are indicating the controller number while the 2 last ones the values. 
Both are split into the MSB and LSB part (7 bits each).
-2 at the end indicates a controller number line

**Reply by unityconstruct** [reply]
Actually a forum _is_ traditionally the place for that _(as evidenced by the original Ctrlr.org forum where this practice frequently yielded 10-20 page threads when numerous people were new & learning - ask me how I know)_. That said, those are bygone times & will be sensitive to the reality that the tenured folks have limited patience with it. Pulled things locally, removed/truncated the worst offenders & will observe the admonition brother 😎

And incidentally, this was EXACTLY what I was trying discern..(msgs working, last token func was unclear) THX!
>I'm completing that by saying that the last attribute -2 or -1 is used to indicate "controller" (-2) or "value" (-1)

**Comment by unityconstruct** [comment, marked as answer]
- `-1,-2` is stored in a token array at index[4]
- function serves as an override for how the message is handled: `NRPN ID` v.s. a `NRPN VALUE`
```
CC,ByteValue,MSB7bitValue,99,-2
CC,ByteValue,LSB7bitValue,98,-2
CC,ByteValue,MSB7bitValue,6,-1
CC,ByteValue,LSB7bitValue,38,-1
```

  - if `-2`
```cpp
bl = MemoryBlock(messageArray.getReference(i).m.getRawData(), messageArray.getReference(i).m.getRawDataSize());
```
  - else
```cpp
bl = midiMessagePattern(messageArray.getReference(i), messageArray.getReference(i).getTokenArray(), getGlobalVariables());
```
## CtrlrMidiMessageEx midiMessageExfromString | CtrlrUtilities.ccp

`ret.overrideValue		= tokens[4].getIntValue();`

```lua
const CtrlrMidiMessageEx midiMessageExfromString (const String &str, const int ch, const int number, const int value)
{
	CtrlrMidiMessageEx ret;
	StringArray tokens;

	tokens.addTokens (str, ",", "\"\'");

	if (tokens.size() >= 5)
	{
		ret.indirectNumberFlag	= indirectFromString (tokens[1]);
		ret.indirectValueFlag	= indirectFromString (tokens[2]);
		ret.overrideNumber		= tokens[3].getIntValue();
		ret.overrideValue		= tokens[4].getIntValue();

		if (tokens[0] == "CC")
			ret.m				= MidiMessage::controllerEvent (ch,number,value);
		if (tokens[0] == "Aftertouch")
			ret.m				= MidiMessage::aftertouchChange (ch,number,value);
		if (tokens[0] == "NoteOn")
			ret.m				= MidiMessage::noteOn (ch,number,(uint8)value);
		if (tokens[0] == "NoteOff")
			ret.m				= MidiMessage::noteOff (ch,number,(uint8)value);
		if (tokens[0] == "ChannelPressure")
			ret.m				= MidiMessage::channelPressureChange (ch,value);
		if (tokens[0] == "ProgramChange")
			ret.m				= MidiMessage::programChange(ch,value);

		if (tokens[0] == "SysEx")
		{
			return (CtrlrSysexProcessor::sysexMessageFromString(tokens[5], value, ch));
		}

		ret.setNumber(number);
		ret.setValue(value);
		return (ret);
	}
	else
	{
	    jassertfalse; // Looks like there is not enough tokens to create a CtrlrMidiMessageEx object
		return (CtrlrMidiMessageEx());
	}
}
```

## CtrlrMidiMessage::patternChanged() : CtrlrMidiMessage.cpp
```lua

void CtrlrMidiMessage::patternChanged()
{
	messagePattern.setSize(0);

	for (int i=0; i<messageArray.size(); i++)
	{
		MemoryBlock bl(0);
		if (messageArray.getReference(i).overrideValue == -2)
		{
			bl = MemoryBlock(messageArray.getReference(i).m.getRawData(), messageArray.getReference(i).m.getRawDataSize());
		}
		else
		{
			bl = midiMessagePattern(messageArray.getReference(i), messageArray.getReference(i).getTokenArray(), getGlobalVariables());
		}

		messagePattern.append (bl.getData(), bl.getSize());
	}
}
```
